File: hedera-mirror-rest-java/src/test/java/com/hedera/mirror/restjava/common/UtilsTest.java
/*
 * Copyright (C) 2024 Hedera Hashgraph, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.hedera.mirror.restjava.common;

import static com.hedera.mirror.restjava.common.ParameterNames.ACCOUNT_ID;
import static com.hedera.mirror.restjava.common.ParameterNames.TOKEN_ID;
import static com.hedera.mirror.restjava.common.Utils.getPaginationLink;
import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.mockito.Mockito.when;

import jakarta.servlet.http.HttpServletRequest;
import java.util.LinkedHashMap;
import java.util.Map;
import org.jetbrains.annotations.NotNull;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.MockedStatic;
import org.mockito.Mockito;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.data.domain.Sort;
import org.springframework.web.context.request.RequestContextHolder;
import org.springframework.web.context.request.ServletRequestAttributes;

@ExtendWith(MockitoExtension.class)
class UtilsTest {

    @Mock
    private HttpServletRequest request;

    @Mock
    private ServletRequestAttributes attributes;

    private MockedStatic<RequestContextHolder> context;

    @BeforeEach
    void setUp() {
        context = Mockito.mockStatic(RequestContextHolder.class);
    }

    @AfterEach
    void closeMocks() {
        context.close();
    }

    @Test
    @DisplayName("Get pagination links")
    void getPaginationLinks() {
        var uri = "/api/v1/1001/accounts/allowances/nfts";
        context.when(RequestContextHolder::getRequestAttributes).thenReturn(attributes);
        when(attributes.getRequest()).thenReturn(request);
        when(request.getRequestURI()).thenReturn(uri);
        when(request.getParameterMap()).thenReturn(Map.of());
        assertThat(getPaginationLink(false, getLastValues(), Sort.Direction.ASC))
                .isEqualTo(uri + "?account.id=gte:0.0.2000&token.id=gt:0.0.6458");
    }

    @Test
    @DisplayName("Get pagination links for all query parameters")
    void getPaginationLinksAllQueryParameters() {
        var uri = "/api/v1/1001/accounts/allowances/nfts";
        Map<String, String[]> params = getParamMap();

        context.when(RequestContextHolder::getRequestAttributes).thenReturn(attributes);
        when(attributes.getRequest()).thenReturn(request);
        when(request.getRequestURI()).thenReturn(uri);
        when(request.getParameterMap()).thenReturn(params);
        assertThat(getPaginationLink(false, getLastValues(), Sort.Direction.ASC))
                .isEqualTo(uri + "?limit=2&order=asc&account.id=gte:0.0.2000&token.id=gt:0.0.6458&owner=false");
    }

    @Test
    @DisplayName("Get pagination links for all query parameters, order descending")
    void getPaginationLinksAllQueryParametersOrderDesc() {
        var uri = "/api/v1/1001/accounts/allowances/nfts";
        Map<String, String[]> params = getParamMap();
        params.put("order", new String[] {"desc"});
        context.when(RequestContextHolder::getRequestAttributes).thenReturn(attributes);
        when(attributes.getRequest()).thenReturn(request);
        when(request.getRequestURI()).thenReturn(uri);
        when(request.getParameterMap()).thenReturn(params);
        assertThat(getPaginationLink(false, getLastValues(), Sort.Direction.DESC))
                .isEqualTo(uri + "?limit=2&order=desc&account.id=lte:0.0.2000&token.id=lt:0.0.6458&owner=false");
    }

    @Test
    @DisplayName("Empty next link")
    void getEmptyPaginationLinks() {
        LinkedHashMap<String, String> lastValues = new LinkedHashMap<>(Map.of(ACCOUNT_ID, "0.0.2000"));
        // When the last element has already been returned
        assertNull(getPaginationLink(true, lastValues, Sort.Direction.ASC));
        // When the lastValues map is null
        assertNull(getPaginationLink(false, null, Sort.Direction.ASC));
        // When the no lastValues have been passed
        assertNull(getPaginationLink(false, new LinkedHashMap<>(Map.of()), Sort.Direction.ASC));
    }

    @Test
    @DisplayName("No request object available")
    void nullRequestObject() {
        context.when(RequestContextHolder::getRequestAttributes).thenReturn(null);
        assertNull(getPaginationLink(false, getLastValues(), Sort.Direction.ASC));
    }

    @NotNull
    private static Map<String, String[]> getParamMap() {
        Map<String, String[]> params = new LinkedHashMap<>();
        params.put("limit", new String[] {"2"});
        params.put("order", new String[] {"asc"});
        params.put("account.id", new String[] {"0.0.1000"});
        params.put("token.id", new String[] {"0.0.1000"});
        params.put("owner", new String[] {"false"});
        return params;
    }

    @NotNull
    private static LinkedHashMap<String, String> getLastValues() {
        LinkedHashMap<String, String> lastValues = new LinkedHashMap<>();
        lastValues.put(ACCOUNT_ID, "0.0.2000");
        lastValues.put(TOKEN_ID, "0.0.6458");
        return lastValues;
    }
}

File: tools/mirror-report/test/utils.test.js
/*
 * Copyright (C) 2024-2025 Hedera Hashgraph, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {parseAccount, toHbars, toIsoString, validateDate} from '../src/utils.js';
import {InvalidArgumentError} from "commander";

describe('parseAccount', () => {
  test('No duplicates', async () => {
    var existing = new Set(['0.0.1']);
    expect(parseAccount('0.0.1', existing)).toEqual(existing);
  });

  test('Multiple', async () => {
    var existing = new Set(['0.0.1']);
    expect(parseAccount('0.0.2', existing)).toEqual(new Set(['0.0.1', '0.0.2']));
  });

  it.each([
    ['0.0.1', ['0.0.1']],
    ['0.0.1-0.0.2', ['0.0.1', '0.0.2']],
    ['0.0.1-0.0.3', ['0.0.1', '0.0.2', '0.0.3']],
  ])("'%s' = %s", (account, expected) => {
    expect(parseAccount(account)).toEqual(new Set(expected));
  });

  it.each([
    null,
    '',
    'x',
    '0-1',
    '100',
    '0.100',
    'x.0.100',
    '0.y.100',
    '0.0.z',
    '0.0.2-0.0.2',
    '0.0.2-0.0.1',
    '0.0.100-0.1.101',
  ])("'%s' throws error", account => {
    expect(() => parseAccount(account)).toThrow(InvalidArgumentError);
  });
});

describe('toHbars', () => {
  it.each([
    [undefined, ''],
    [null, ''],
    [0n, '0.0'],
    [1n, '0.00000001'],
    [-1n, '-0.00000001'],
    [10n, '0.00000010'],
    [99999999n, '0.99999999'],
    [-99999999n, '-0.99999999'],
    [100000000n, '1.0'],
    [-100000000n, '-1.0'],
    [10000000001n, '100.00000001'],
    [4999999999999999999n, '49999999999.99999999'],
    [-4999999999999999999n, '-49999999999.99999999'],
    [5000000000000000000n, '50000000000.0']
  ])("toHbars(%s) = '%s'", (tinybars, expected) => {
    expect(toHbars(tinybars)).toBe(expected);
  });
});

describe('toIsoString', () => {
  it.each([
    [undefined, ''],
    [null, ''],
    ['0.0', '1970-01-01T00:00:00.000000000Z'],
    ['0.', '1970-01-01T00:00:00.000000000Z'],
    ['0.000000001', '1970-01-01T00:00:00.000000001Z'],
    ['1.000000001', '1970-01-01T00:00:01.000000001Z'],
    ['1.1', '1970-01-01T00:00:01.000000001Z'],
    ['1733181006', '2024-12-02T23:10:06.000000000Z'],
    ['1733181006.483579814', '2024-12-02T23:10:06.483579814Z']
  ])("toIsoString(%s) = '%s'", (timestamp, expected) => {
    expect(toIsoString(timestamp)).toBe(expected);
  });
});

describe('validateDate', () => {
  test('Valid', async () => {
    const date = '2024-12-17';
    expect(validateDate(date)).toBe(date);
  });

  it.each([
    null,
    '',
    '1970-01-01T00:00:00.000000000Z',
    '19700101',
    '1970-13-01',
    '1970-12-32',
    '99999-01-01',
    '999-01-01',
  ])("'%s' throws error", date => {
    expect(() => validateDate(date)).toThrow(InvalidArgumentError);
  });
});
